klippy/gcode.py

```diff
@@ -349,6 +360,24 @@ def _process_data(self, eventtime):
         if self.fd_handle is None:
             self.fd_handle = self.reactor.register_fd(self.fd,
                                                       self._process_data)
+    def _handle_remote_help(self, web_request):
+        web_request.send(dict(self.gcode_help))
+    def _handle_remote_restart(self, web_request):
+        path = web_request.get_path()
+        if path == "gcode/restart":
+            web_request.put('script', 'restart')
+        elif path == "gcode/firmware_restart":
+            web_request.put('script', 'firmware_restart')
+        self.run_script_from_remote(web_request)
+    def run_script_from_remote(self, web_request):
+        script = web_request.get('script')
+        if 'M112' in script.upper():
+            self.cmd_M112({})
+            return
```

Comment:
I'd recommend against checking for an asynchronous M112 in the g-code scripts.  That feature of g-code is an ugly hack that is painful to work with.  I'd think we could instead introduce an "emergency stop" endpoint that we could have confidence would run asynchronously.

Reply:
Sounds good.  Presumably the "emergency stop" endpoint's handler would simply run `gcode.cmd_M112({})`?

Reply:
That sounds good.  (Or, `gcode.cmd_M112(None)` or we can make it explicit that cmd_M112 doesn't need a parameter.)